Framework/Algorithms/src/CropToComponent.cpp

```diff
@@ -0,0 +1,129 @@
+
+namespace {
+
+void getDetectors(Mantid::API::MatrixWorkspace_sptr workspace, const std::vector<std::string>& componentNames, std::vector<Mantid::Geometry::IDetector_const_sptr>& detectors) {
+  auto instrument = workspace->getInstrument();
+  for (const auto &componentName : componentNames) {
+      instrument->getDetectorsInBank(detectors, componentName);
+  }
+}
+
+void getDetectorIDs(std::vector<Mantid::Geometry::IDetector_const_sptr>&detectors, std::vector<Mantid::detid_t>& detectorIDs){
+  auto numberOfDetectors = detectors.size();
+  for (size_t index = 0; index < numberOfDetectors; ++index) {
+      auto det = detectors[index];
+      detectorIDs[index] = det->getID();
+  }
+}
+}
+
+
+namespace Mantid
+{
+namespace Algorithms
+{
+
+using Mantid::Kernel::Direction;
+using Mantid::API::WorkspaceProperty;
+
+// Register the algorithm into the AlgorithmFactory
+DECLARE_ALGORITHM(CropToComponent)
+
+//----------------------------------------------------------------------------------------------
+
+/// Algorithms name for identification. @see Algorithm::name
+const std::string CropToComponent::name() const { return "CropToComponent"; }
+
+/// Algorithm's version for identification. @see Algorithm::version
+int CropToComponent::version() const { return 1; }
+
+/// Algorithm's category for identification. @see Algorithm::category
+const std::string CropToComponent::category() const
+{
+    return "TODO: FILL IN A CATEGORY";
+}
+
+/// Algorithm's summary for use in the GUI and help. @see Algorithm::summary
+const std::string CropToComponent::summary() const
+{
+    return "Crops a workspace to a set of components.";
+}
+
+//----------------------------------------------------------------------------------------------
+/** Initialize the algorithm's properties.
+ */
+void CropToComponent::init()
+{
+    declareProperty(Kernel::make_unique<WorkspaceProperty<Mantid::API::MatrixWorkspace>>(
+                        "InputWorkspace", "", Direction::Input),
+                    "An input workspace.");
+    declareProperty(Kernel::make_unique<WorkspaceProperty<Mantid::API::MatrixWorkspace>>(
+                        "OutputWorkspace", "", Direction::Output),
+                    "An output workspace.");
+    declareProperty(
+        Kernel::make_unique<Mantid::Kernel::ArrayProperty<std::string>>(
+            "ComponentNames"),
+        "List of component names to which the workspace is goign to be cropped "
+        "to.");
+}
+
+//----------------------------------------------------------------------------------------------
+/** Execute the algorithm.
+ */
+void CropToComponent::exec()
+{
+    // Get the names of the components
+    std::vector<std::string> componentNames = getProperty("ComponentNames");
+    Mantid::API::MatrixWorkspace_sptr inputWorkspace
+        = getProperty("InputWorkspace");
+    Mantid::API::MatrixWorkspace_sptr outputWorkspace
+        = getProperty("OutputWorkspace");
+
+    // Check if the component names are empty
```

Comment:
Should comment say, something like, // If component names are empty, make output workspace the same as input workspace. ?